apps/condo/domains/contact/tasks/exportContacts.js

```diff
@@ -0,0 +1,166 @@
+            id,
+        },
+    }
+}
+
+async function exportContacts (taskId) {
+    if (!taskId) throw new Error('taskId is required')
+    const { keystone: context } = await getSchemaCtx('ContactExportTask')
+
+    const task = await ContactExportTask.getOne(context, { id: taskId })
+    const { format, where, sortBy } = task
+
+    const baseAttrs = { dv: 1, sender: TASK_DV_AND_SENDER }
+
+    if (!task.locale) {
+        await ContactExportTask.update(context, taskId, {
+            ...baseAttrs,
+            status: ERROR,
+        })
+
+        throw new Error(`ContactExportTask with id = ${taskId} doesn't have value for "locale" field`)
+    }
+
+    try {
+        setLocaleForKeystoneContext(context, task.locale)
+
+        const allRoles = await ContactRole.getAll(context, {
+            deletedAt: null,
+            OR: [
+                { organization_is_null: true },
+                { organization: { id: where.organization.id } },
+            ],
+        })
+        const translatedRolesMap = Object.fromEntries(allRoles.map(role => ([role.id, role.name])))
+
+        const contactLoader = await buildContactsLoader({ where, sortBy })
+        const totalRecordsCount = await Contact.count(context, where)
+
+        const loadRecordsBatch = async (offset, limit) => {
+            const contacts = await contactLoader.loadChunk(offset, limit)
+
+            this.progress(Math.floor(offset / totalRecordsCount * 100))
+
+            return contacts
+        }
+
+        const convertRecordToFileRow = (contact) => contactToRow({
+            task, contact, translatedRolesMap,
+        })
+
+        switch (format) {
+            case EXCEL: {
+                if (totalRecordsCount > MAX_XLSX_FILE_ROWS) {
+                    await exportRecordsAsCsvFile({
+                        context,
+                        loadRecordsBatch,
+                        convertRecordToFileRow,
+                        baseAttrs,
+                        taskServerUtils: ContactExportTask,
+                        totalRecordsCount,
+                        taskId,
+                    })
+                } else {
+                    await exportRecordsAsXlsxFile({
+                        context,
+                        loadRecordsBatch,
+                        convertRecordToFileRow,
+                        buildExportFile: (rows) => buildExportFile({ rows, task }),
+                        baseAttrs,
+                        taskServerUtils: ContactExportTask,
+                        totalRecordsCount,
+                        taskId,
+                    })
+                }
+                break
+            }
+        }
+    } catch (error) {
+        taskLogger.error({
+            msg: 'Failed to export contacts',
+            data: { id: taskId },
+            error,
+        })
+        throw error
+    }
+}
+
+module.exports = {
+    exportContacts: createTask('exportContacts', exportContacts, { priority: 2 }),
```

Comment:
It’s better to use default priority

Reply:
It is equal to the same solution at ticket domain